flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java

```diff
@@ -213,6 +218,23 @@ public boolean discard(JobStatus jobStatus) throws Exception {
 
 	void discard() throws Exception {
 		try {
+			// check if we have a shortcut hook to dispose the state
+			final StateDisposeHook disposeHook = this.disposeHook;
+			if (disposeHook != null) {
+				// fast path!
```

Comment:
On a second look, do you think it could even make sense to have the previous behaviour also encapsulated as a cleanup hook and get rid of the if completely?